src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ReliableMessagingHelpers.cs

```diff
@@ -2,11 +2,215 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Globalization;
 using System.Runtime;
+using System.ServiceModel.Security;
+using System.Threading;
 using System.Threading.Tasks;
+using System.Xml;
 
 namespace System.ServiceModel.Channels
 {
+    internal delegate Task OperationWithTimeoutAsyncCallback(TimeSpan timeout);
+
+    internal sealed class Guard
+    {
+        private TaskCompletionSource<object> _tcs;
+        private int _currentCount = 0;
+        private int _maxCount;
+        private bool _closed;
+        private object _thisLock = new object();
+
+        public Guard() : this(1) { }
+
+        public Guard(int maxCount)
+        {
+            _maxCount = maxCount;
+        }
+
+        public void Abort()
+        {
+            _closed = true;
+        }
+
+        public async Task CloseAsync(TimeSpan timeout)
+        {
+            lock (_thisLock)
+            {
+                if (_closed)
+                    return;
+
+                _closed = true;
+
+                if (_currentCount > 0)
+                    _tcs = new TaskCompletionSource<object>();
```

Comment:
You probably want to create the TaskCompletionSource with TaskCreationOptions.RunContinuationsAsynchronously. That will cause the continuation to be queued to the thread pool. Otherwise, you are going to execute the completion within the call to SetResult, which is almost certainly not what you want here.

DO NOT use the similarly named TaskContinuationOptions.RunContinuationsAsynchronously.

Reply:
I think I've hit that bug like 3 times

Reply:
I half-seriously think we should add a TaskCompletionSource2 whose only job is to default to passing TaskCreationOptions.RunContinuationsAsynchronously, so you don't either forget it or mistakenly do TaskContinuationOptions instead...

We actually added an FxCop rule to detect when you use TaskContinuationOptions.RunContinuationsAsynchronously.

Reply:
You're right, this is needed here.

Reply:
I was trying to work out how you can even pass a TaskContinuationOptions to the constructor, then it clicked and I realized it gets boxed and passed as the state object. TaskCompletionSource isn't sealed so you could derive it in your own code base and use that.